src/MongoDB.Driver.Core/Core/Bindings/ServerSelectionHelper.cs

```diff
@@ -0,0 +1,88 @@
+/* Copyright 2019–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Bindings;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Servers;
+
+namespace MongoDB.Driver.Core.Core.Bindings
+{
+    /// <summary>
+    /// This class contains helper methods for server selection with sharded transactions.
+    /// </summary>
+    internal static class ServerSelectionHelper
+    {
+        public static IServer SelectServerAndPinIfNeeded(
+            ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            var server = SelectServer(cluster, session, selector, cancellationToken);
+            PinShardRouterToSessionIfNeeded(cluster, session, server);
+            return server;
```

Comment:
I feel that the two methods are actually less easy to understand because you've spread the pinning logic across two different places (methods). Half the pinning logic is related to selecting a server and the other half is related to remembering that we are pinned. In the above suggestion both halves of the logic are together.

I certainly grok how your code works, I just think it's unnecessarily complicated and indirect.